flink-runtime/src/test/java/org/apache/flink/runtime/concurrent/FutureUtilsTest.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.concurrent;
+
+import org.apache.flink.runtime.concurrent.FutureUtils.ConjunctFuture;
+import org.apache.flink.runtime.concurrent.impl.FlinkCompletableFuture;
+
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.concurrent.ExecutionException;
+
+import static org.junit.Assert.*;
+
+/**
+ * Tests for the utility methods in {@link FutureUtils}
+ */
+public class FutureUtilsTest {
+
+	@Test
+	public void testConjunctFutureFailsOnEmptyAndNull() throws Exception {
+		try {
+			FutureUtils.combineAll(null);
+			fail();
+		} catch (NullPointerException ignored) {}
+
+		try {
+			FutureUtils.combineAll(Collections.<Future<?>>emptyList());
+			fail();
+		} catch (IllegalArgumentException ignored) {}
+
+		try {
+			FutureUtils.combineAll(Arrays.asList(
+					new FlinkCompletableFuture<Object>(),
+					null,
+					new FlinkCompletableFuture<Object>()));
+			fail();
+		} catch (NullPointerException ignored) {}
+	}
+
+	@Test
+	public void testConjunctFutureCompletion() throws Exception {
+		// some futures that we combine
+		CompletableFuture<Object> future1 = new FlinkCompletableFuture<>();
+		CompletableFuture<Object> future2 = new FlinkCompletableFuture<>();
+		CompletableFuture<Object> future3 = new FlinkCompletableFuture<>();
+		CompletableFuture<Object> future4 = new FlinkCompletableFuture<>();
+
+		// some future is initially completed
+		future2.complete(new Object());
+
+		// build the conjunct future
+		ConjunctFuture result = FutureUtils.combineAll(Arrays.asList(future1, future2, future3, future4));
+
+		Future<Void> resultMapped = result.thenAccept(new AcceptFunction<Void>() {
+			@Override
+			public void accept(Void value) {}
+		});
```

Comment:
This is simply added as a test that applyied functions (and their result futures) are not called before completion.